crates/pathfinder/src/state/sync/l2.rs

```diff
@@ -46,13 +49,72 @@ pub enum Event {
     /// The receiver should return true (if the contract exists) or false (if it does not exist)
     /// for each contract using the [oneshot::channel].
     QueryContractExistance(Vec<ClassHash>, oneshot::Sender<Vec<bool>>),
+    /// A new L2 pending update was polled.
+    Pending(Box<PendingBlock>, Box<sequencer::reply::StateUpdate>),
+}
+
+/// Poll's the Sequencer's pending block and emits [Event::Pending]
+/// until the pending block is no longer connected to our current head.
+///
+/// This disconnect is detected whenever
+/// - `pending.block_hash != head`, or
+/// - `pending` is a fully formed block and not [PendingBlock]
+async fn poll_pending(
+    tx_event: &mpsc::Sender<Event>,
+    sequencer: &impl sequencer::ClientApi,
+    head: StarknetBlockHash,
+    poll_interval: Duration,
+) -> anyhow::Result<()> {
+    use crate::core::BlockId;
+
+    loop {
+        use sequencer::reply::MaybePendingBlock;
+
+        let block = match sequencer
+            .block(BlockId::Pending)
+            .await
+            .context("Download block")?
+        {
+            MaybePendingBlock::Block(block) => {
+                tracing::debug!(hash=%block.block_hash, "Found full block, exiting pending mode.");
+                return Ok(());
+            }
+            MaybePendingBlock::Pending(pending) if pending.parent_hash != head => {
+                tracing::debug!(
+                    pending=%pending.parent_hash, head=%head,
+                    "Pending block's parent hash does not match head, exiting pending mode"
+                );
+                return Ok(());
+            }
+            MaybePendingBlock::Pending(pending) => pending,
+        };
+
+        // Download the pending state diff.
+        let state_update = sequencer
+            .state_update(BlockId::Pending)
+            .await
+            .context("Download state update")?;
+        if state_update.block_hash.is_some() {
+            tracing::debug!("Found full state update, exiting pending mode.");
+            return Ok(());
+        }
```

Comment:
Similarly to the above - if this is a valid state update - can we just re-use it instead of re-downloading?